src/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -13,17 +13,25 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
         {
             if (source == null) throw Error.ArgumentNull("source");
             IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null) return partition.ElementAt(index);
-            IList<TSource> list = source as IList<TSource>;
-            if (list != null) return list[index];
-            if (index >= 0)
+            if (partition != null)
+            {
+                bool found;
+                TSource element = partition.TryGetElementAt(index, out found);
+                if (found) return element;
+            }
+            else
             {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                IList<TSource> list = source as IList<TSource>;
+                if (list != null) return list[index];
+                if (index >= 0)
                 {
-                    while (e.MoveNext())
+                    using (IEnumerator<TSource> e = source.GetEnumerator())
                     {
-                        if (index == 0) return e.Current;
-                        index--;
+                        while (e.MoveNext())
+                        {
+                            if (index == 0) return e.Current;
```

Comment:
I did such experiment in the past. It was faster. In some cases pretty noticeably. I think that was mostly due to not having to store intermediate results in fields.

That was however not in this context. I had a completely alternative LINQ implementation based on ISequence/Iterator with T TryMoveNext(out bool success) signature + C# support for foreach and ISequence iterators. 
The overall thing was a fun experiment but many years too late to the world dominated by IEnumerable :-) 

Not sure if added costs of bridging into IEnumerable would be prohibiting for using this pattern internally in Iterator.